Workflow changes 0.5

hbooth opened the pull request:
Adding in a description of Paging and the Request workflow and then updating all of the various endpoints to use this. Also added a few new endpoints to better facilitate metadata creation. A number of requests have been updated so please take a look and make sure you are ok with this.

There is a still an open issue with the specification for the algorithm prerequisites that will need to be improved upon, but that is fairly minor relative to the rest of these changes and I didn't want to hold this up any longer than necessary given the limited time available.

smuellerDD commented:
Am Freitag, 29. März 2019, 22:42:21 CET schrieb Harold Booth:

Hi Harold,

Some comments:

- 10.5.1 paging: please add some words how the incomplete property is added to 
responses - I see sometimes a reference to paging in subsequent sections and 
sometimes not. Is paging only applicable for those endpoints where it was 
mentioned or in general to all endpoints?

- 10.5.2: please clarify whether this response format is only due to a 
specific paged request or whether all regular (unpaged) requests now follow 
that format.

- 10.7.3.1: url and contactUrl not defined

- how does 10.8 relate to the address array of 10.7.2.1?

- how does 10.9 relate to the email/phone number of 10.7.2.1?

- 10.14.4.1: signature is not explained - which key to use, which algo to use, 
what to sign


In general: I remember you mentioned that you want to provide a process how 
this meta data is suggested to be used to prevent cluttering the database with 
duplicate entries. Do you happen to have such process already?

Ciao
Stephan

hbooth commented:
> 10.5.1 paging: please add some words how the incomplete property is added to responses - I see sometimes a reference to paging in subsequent sections and sometimes not. Is paging only applicable for those endpoints where it was mentioned or in general to all endpoints? - 10.5.2: please clarify whether this response format is only due to a specific paged request or whether all regular (unpaged) requests now follow that format.

Added a clarification to indicate that paging is only on those resource operations which indicate it in the specification. Primarily the operations which returns lists of items are the ones which I have tried to indicate this for. If I missed any let me know.

> - 10.7.3.1: url and contactUrl not defined

It is defined in 10.7. Did you miss it?

> how does 10.8 relate to the address array of 10.7.2.1?

Added clarification to the document.

> how does 10.9 relate to the email/phone number of 10.7.2.1? 

Added clarification to the document.

> 10.14.4.1: signature is not explained - which key to use, which algo to use, what to sign In general:

Currently, this intentionally not explained with my update. Please wait on this. I don't think this information is necessary but need to verify. Once I get confirmation, I intend to remove it.

> I remember you mentioned that you want to provide a process how this meta data is suggested to be used to prevent cluttering the database with duplicate entries. Do you happen to have such process already?

So, we have the beginnings of things internally to manage normalization, although it's somewhat manual at the present time, we do plan to improve on it. I would like to expose some ability to have some search query parameters on the various "listing" endpoints to make it possible for external users to locate existing data as well. I have a lot of the searching already implemented internally too but it would require a little bit more time to document this as well. I may try to add it in later in the week, but no promises.

I believe the above answers your question but let me know.

smuellerDD commented:
Am Sonntag, 31. März 2019, 03:56:59 CEST schrieb Harold Booth:

Hi Harold,

> > - 10.7.3.1: url and contactUrl not defined
> 
> It is defined in 10.7. Did you miss it?

Well, I have seen the wording but it is not readily clear.

contactUrl - do you want to say that it contains the link to the person? If 
so, please use the name keywords and endpoints. I.e. either /person and 
personUrl or /contact and contactUrl. 

The same with organizationUrl and the reference to a vendor. I.e. either 
organizationUrl and /organization or vendorUrl and /vendor.

Ciao
Stephan

smuellerDD commented:
Am Freitag, 29. März 2019, 22:42:21 CEST schrieb Harold Booth:

Hi Harold,

10.10 defines that a module must be combined with an addressUrl. However, when 
using a "simple" workflow approach where one only registers a vendor with 
10.7.2 including an address definition, how would I retrieve the addressUrl? I 
guess you do not want us to iterate over the address end point to only find 
out our address ID.

Ciao
Stephan

hbooth commented:
> contactUrl - do you want to say that it contains the link to the person? If so, please use the name keywords and endpoints. I.e. either /person and personUrl or /contact and contactUrl. 

The contactsUrl  follows the form of `/vendor/{vendorId}/contacts`, and returns a list of person resources who are all associated with the vendor. I can change this to persons, but my concern is that I did not want to confuse this endpoint with the /persons endpoint, which is independent of any specific vendor.

> The same with organizationUrl and the reference to a vendor. I.e. either organizationUrl and /organization or vendorUrl and /vendor.

Agreed, I can make them the same for this one. We call it organizations internally, but the ACVP has used the name vendors. Does anyone have any preferences? I have a preference for using organization throughout in lieu of vendor as organization is more generic.

At the current time, I have made it vendorUrl just to keep it in-line with current naming. Switching in spec and code is trivial, but a consensus will be necessary as soon as possible.

hbooth commented:
Updated documentation to include a `url` property on each address element within the addresses array. The code was already doing this, but I forgot to add it to the documentation. This helps somewhat with the question below.

> 10.10 defines that a module must be combined with an addressUrl. However, when using a "simple" workflow approach where one only registers a vendor with 10.7.2 including an address definition, how would I retrieve the addressUrl?

If you have a `vendorUrl`, you can either retrieve the `url` of the address resource for the address you want in the `addresses` array on 10.7.3.1 or from the `/vendors/{vendorId}/addresses` resource described in 10.8.1.

If you are creating a vendor from scratch you will need to use 10.7.2 and then retrieve the `approvedUrl` using 10.6 and then following one of the approaches above.

smuellerDD commented:
Am Sonntag, 31. März 2019, 23:26:52 CEST schrieb Harold Booth:

Hi Harold,

Thank you. I have no preference.

Ciao
Stephan

smuellerDD commented:
Am Sonntag, 31. März 2019, 23:53:30 CEST schrieb Harold Booth:

Hi Harold,

> If you are creating a vendor from scratch you will need to use 10.7.2 and
> then retrieve the `approvedUrl` using 10.6 and then following one of the
> approaches above.

That is an interesting one: As of now I have neglected 10.6 because I have 
seen no link from other sections to this one. Can you please add some more 
clarifications on how 10.6 ties into the other operations?

I.e.: How would I get the request URL to follow up the request? What type of 
binding exists between a particular request and the requested URL?

Ciao
Stephan

smuellerDD commented:
Am Montag, 1. April 2019, 07:21:13 CEST schrieb Stephan Müller:

Hi,

> Am Sonntag, 31. März 2019, 23:53:30 CEST schrieb Harold Booth:
> 
> Hi Harold,
> 
> > If you are creating a vendor from scratch you will need to use 10.7.2 and
> > then retrieve the `approvedUrl` using 10.6 and then following one of the
> > approaches above.
> 
> That is an interesting one: As of now I have neglected 10.6 because I have
> seen no link from other sections to this one. Can you please add some more
> clarifications on how 10.6 ties into the other operations?

Sorry for the noise, I see it. Please disregard this request.

Ciao
Stephan

smuellerDD commented:
Am Montag, 1. April 2019, 07:21:13 CEST schrieb Stephan Müller:

Hi Harold,

> Am Sonntag, 31. März 2019, 23:53:30 CEST schrieb Harold Booth:
> 
> Hi Harold,
> 
> > If you are creating a vendor from scratch you will need to use 10.7.2 and
> > then retrieve the `approvedUrl` using 10.6 and then following one of the
> > approaches above.

For 10.6, how is the expected workflow for the status indicator? I.e. what 
should I do if I receive anything other than "accepted"?


Ciao
Stephan

smuellerDD commented:
Am Sonntag, 31. März 2019, 23:53:30 CEST schrieb Harold Booth:

Hi Harold,

> If you are creating a vendor from scratch you will need to use 10.7.2 and
> then retrieve the `approvedUrl` using 10.6 and then following one of the
> approaches above.

How exactly is that going to work? 10.6 seems to indicate that only one URL 
can be returned. But 10.7.3 seems to contain two URLs that are returned (url 
and contactUrl).

I would guess it is just url that is returned with 10.6 and I need to make a 
query to that URL to get the contact URL. If that is the case, could you add 
that detail about which endpoint 10.6 returns to 10.7.2.2?

Ciao
Stephan

smuellerDD commented:
Am Freitag, 29. März 2019, 22:42:21 CEST schrieb Harold Booth:

Hi Harold,

- Is 10.12.3 pre-defined by NIST? If so, how would that list be updated?

- 10.12: How would we relay the information which processor features were used 
(RDRAND, etc)?

Ciao
Stephan

hbooth commented:
> If that is the case, could you add that detail about which endpoint 10.6 returns to 10.7.2.2?

Additional descriptive text was added to each type.

hbooth commented:
> Is 10.12.3 pre-defined by NIST? If so, how would that list be updated? - 10.12: How would we relay the information which processor features were used (RDRAND, etc)?

This particular area is not well-developed at the moment. I was hoping the various CMUF working groups operating in this space would be further along than they currently are, which would have provided more clarity on what is needed here. In the meantime, there is no checking on the property/value pairs to insure the properties actually exists. We will rely on good faith efforts to characterize their OEs appropriately using the mechanisms in place and evolve the requirements as needs dictate. To do RDRAND specifically, I would add something like `rdrand` for the name and `required` for the value or something like that. In the future an array value for a property name of `features` may make more sense, but we don't have this capability built-in at this time.

smuellerDD commented:
Am Montag, 1. April 2019, 18:23:55 CEST schrieb Harold Booth:

Hi Harold,

So you are expecting that vendors update that list? I think I am missing the 
POST/PUT definitions for this endpoint.

Ciao
Stephan

hbooth commented:
> So you are expecting that vendors update that list? I think I am missing the POST/PUT definitions for this endpoint.

There is no expectation that anyone will use the `GET /dependencies/properties` endpoint at this moment. (It's marked as an optional endpoint for the spec right now as well.) Looking forward, as we understand the requirements we do intend to add on to this and improve this endpoint to better address the requirements surrounding the specification of Operation Environments.

The creation of dependencies will work for any arbitrary name/value pair and the expectation is that submitters will provide enough information using this mechanism to at least describe the environment and then as we review the name/value pairs chosen, we would like to normalize and standardize around the commonly used ones.
